### ISSUE TITLE
Regression on FeedSpam/Ham Task

### ISSUE THREAD

**Issue by quantranhong1999** [linagora/james-project#4658]
### When
When running Feed Spam/Ham Tasks on TMail pre-prod (0.6.4 rc1) against normal Rspamd (not per-user Bayes enabled).

### Expected
The messages within the input `period` are reported.

### Actual behavior
- All the messages are counted regardless of `period`
- No message is reported to Rspamd (no error log on either James or Rspamd, no learned metric increased...)

Sample response:
```json
{
    "additionalInformation": {
        "type": "FeedHamToRspamdTask",
        "timestamp": "2022-10-26T06:14:44.404815997Z",
        "hamMessageCount": 73717583,
        "reportedHamMessageCount": 0,
        "errorCount": 0,
        "runningOptions": {
            "periodInSecond": 3600,
            "messagesPerSecond": 10,
            "samplingProbability": 1.0,
            "classifiedAsSpam": null,
            "rspamdTimeoutInSeconds": 15
        }
    },
    "type": "FeedHamToRspamdTask",
    "submittedFrom": "james-jmap-85c96858c8-8ndvk",
    "executedOn": "james-imap-smtp-7bcf5bb994-pgvx2",
    "cancelledFrom": null,
    "submitDate": "2022-10-26T05:56:17.681+0000",
    "status": "completed",
    "taskId": "1ff5be9c-64db-4588-9132-3e9d98991457",
    "completedDate": "2022-10-26T06:14:44.419+0000",
    "startedDate": "2022-10-26T05:56:17.961+0000"
}
```

rf: https://james-admin.upn.integration-open-paas.org/tasks/1ff5be9c-64db-4588-9132-3e9d98991457

These tasks have worked normally on Tung's reactor timeout work (~ 0.6.3 release), we likely need to investigate what changed since then.

**Comment by quantranhong1999** [comment]
0.6.3.1 has the same issue. The weird thing is its latest James commit is `33ac4d8167bc3a4f26bd988105dff201416fe3e4` which is even before the Tung timeout commit while Tung tested normally before... I doubt there are some parallel PR/commits that led to this situation.

**Comment by chibenwa** [comment]
Write a unit test and debug ;-)

**Comment by quantranhong1999** [comment]
Reproduced here: https://github.com/apache/james-project/pull/1276/commits/351afd76f860b7651344107df4887cee52747b9b

### TLDR
migration issue (null internalDate on preprod), not code bug

### What happened?
messages with null internalDate -> NP/Predicate check exception (not sure what but let's say RuntimeException) and errors are skipped thanks to Throwing.function/Reactor throttle
  -> `context.incrementHamMessageCount()` still increases while all null internalDate messages are filtered out/skipped (that is why no increase for `reportedCount` or `errorCount`) -> fall into `.switchIfEmpty(Mono.just(Result.COMPLETED))` at the end.

#### Why does it seem normal with `null`/big period but not for the small period?

```java
.filter(mailboxMessageMetaData -> afterDate.map(date -> mailboxMessageMetaData.getInternalDate().after(date)).orElse(true))
```
- **null period** -> always returns true without checking against null internalDate
- **big period** -> reactor lazy evaluation (downstream operators report messages to Rspamd and slow down the evaluation of upstream message count)
- **small period** -> likely there are no new messages in that period and only null internalDate messages -> downstream operator quickly skipped (exception) lead to quick upstream message counting.

**Comment by quantranhong1999** [comment]
Forget about the null internalDate, I think it is just simply that there are no spam messages on the last 1 day or last 1-week cause likely the last messages were from 19/10.

Can not send new mail to check right now (issue reported on signal), but I could check within the last 19/10 messages range and the feed spam task counts/reports them exactly (e.g report two more spam messages then the task counts/reports 2 more messages).

**So I think this is not a bug of Tasks but the rationale is we did not able to send any new mail on pre-prod**

**However, I propose some improvements for feed spam/ham tasks**:
- Just count the messages in period instead count all (as I committed [here](https://github.com/apache/james-project/pull/1276/commits/b7df6a6fd48af09937da2a45145e19bbcf201737))
- Improve edge cases:
  - When the user moves an old message (say 3 days ago) to the Spam mailbox, he would want that message to be learned as Spam.
    Today the moved message inherits the old `internalDate` -> James can omit them when running tasks with small periods like 1 day.
  - User would want messages in sub mailboxes of Spam (e.g Spam.subMailbox1) to be learned as Spam
    Today we do exactly compare mailbox names.
  - User would not want messages in sub mailboxes of Trash (e.g Trash.sub1) to be learned as Ham
    Today we do exactly compare mailbox names.

**Comment by quantranhong1999** [comment]
> Just count the messages in period instead count all

https://github.com/apache/james-project/pull/1284
